# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1314059247 14400
#      Mon Aug 22 20:27:27 2011 -0400
# Branch stable
# Node ID 7626f893446639b7997918f1adab54a1d6e2626a
# Parent  9f3ddb1ff2e307a788dc004c84e89b7596d5f5e3
correctly resize scanf output for %c formats (bug #34037)

* oct-stream.cc (BEGIN_C_CONVERSION): Resize tmp buffer to number of
characters read.

diff --git a/src/oct-stream.cc b/src/oct-stream.cc
--- a/src/oct-stream.cc
+++ b/src/oct-stream.cc
@@ -1516,17 +1516,19 @@ do_scanf_conv (std::istream&, const scan
  \
   int c = EOF; \
   int n = 0; \
  \
   while (is && n < width && (c = is.get ()) != EOF) \
     tmp[n++] = static_cast<char> (c); \
  \
   if (n > 0 && c == EOF) \
-    is.clear ()
+    is.clear (); \
+ \
+  tmp.resize (n)
 
 // For a `%s' format, skip initial whitespace and then read until the
 // next whitespace character or until WIDTH characters have been read.
 #define BEGIN_S_CONVERSION() \
   int width = elt->width; \
  \
   std::string tmp; \
  \
diff --git a/test/test_io.m b/test/test_io.m
--- a/test/test_io.m
+++ b/test/test_io.m
@@ -237,16 +237,19 @@
 %!assert(puts (1),-1);
 
 %% test/octave.test/io/puts-3.m
 %!error <Invalid call to puts.*> puts ();
 
 %% test/octave.test/io/puts-4.m
 %!error <Invalid call to puts.*> puts (1, 2);
 
+%!assert (sscanf ('123456', '%10c'), '123456')
+%!assert (sscanf ('123456', '%10s'), '123456')
+
 %% test/octave.test/io/sscanf-1.m
 %!test
 %! [a, b, c] = sscanf ("1.2 3 foo", "%f%d%s", "C");
 %! [v1, c1, m1] = sscanf ("1 2 3 4 5 6", "%d");
 %! [v2, c2, m2] = sscanf ("1 2 bar 3 4 5 6", "%d");
 %! 
 %! assert((a == 1.2 && b == 3 && c == "foo"
 %! && v1 == [1; 2; 3; 4; 5; 6] && c1 == 6 && ischar (m1)
